qiskit/transpiler/passes/optimize_phase_shift_gates.py

```diff
@@ -0,0 +1,105 @@
+
+        Args:
+            dag (DAGCircuit): the directed acyclic graph to run on.
+        Returns:
+            DAGCircuit: Transformed DAG.
+        """
+
+        n = dag.width()  # number of qubits in the circuit
+        T_counter = {}  # Counts the cumulative phase of T-gates and other z-rotation gates
+        T_position_counter = {}  # Keeps track of the locations of T-gates and other z-rot. gates
+        state_tracker = np.zeros((dag.width(), dag.size())).astype('int')
+        for i in range(dag.width()):
+            state_tracker[i][i] = 1 # initial state
+        k = dag.width()  # number of initial variables
+        for e in dag.topological_nodes():
+            if e.data_dict['type'] == 'op':
+                if e.data_dict['name'] == 'cx':
+                    state_tracker[e.data_dict['qargs'][1][1]] ^= state_tracker[e.data_dict['qargs'][0][1]]
+                elif e.data_dict['name'] in ['t', 'tdg', 's', 'sdg', 'z']:
+                    phase_T_gate_equivalent = {'t': 1, 'tdg': 7, 's': 2, 'sdg': 6, 'z': 4}[e.data_dict['name']]
+                    if str(state_tracker[e.data_dict['qargs'][0][1]]) in T_counter:
+                        T_counter[str(state_tracker[e.data_dict['qargs'][0][1]])] += phase_T_gate_equivalent
+                        T_position_counter[str(state_tracker[e.data_dict['qargs'][0][1]])].append(e._node_id)
+                    else:
+                        T_counter[str(state_tracker[e.data_dict['qargs'][0][1]])] = phase_T_gate_equivalent
+                        T_position_counter[str(state_tracker[e.data_dict['qargs'][0][1]])] = [e._node_id]
+                elif e.data_dict['name'] == 'u1':
+                    if str(state_tracker[e.data_dict['qargs'][0][1]]) in T_counter:
+                        T_counter[str(state_tracker[e.data_dict['qargs'][0][1]])] += e.data_dict['op'].params[0] * 4/np.pi
+                        T_position_counter[str(state_tracker[e.data_dict['qargs'][0][1]])].append(e._node_id)
+                    else:
+                        T_counter[str(state_tracker[e.data_dict['qargs'][0][1]])] = e.data_dict['op'].params[0] * 4/np.pi
+                        T_position_counter[str(state_tracker[e.data_dict['qargs'][0][1]])] = [e._node_id]
+                else:
+                    state_tracker[e.data_dict['qargs'][0][1]] = np.zeros(dag.size())
+                    state_tracker[e.data_dict['qargs'][0][1]][k] = 1
+                    k += 1
+        circuit_length_min = len(dag.multi_graph.nodes) + 1
+        circuit_depth_min = dag.depth() + 1
+        for key in T_counter:
+            if T_counter[key] % 8 == 0:
+                for i in T_position_counter[key]:
+                    dag.remove_op_node(i)  # Remove all the nodes that cancelled out
+            else:
+                original_dag = copy.deepcopy(dag)
+                for final_gate_position in range(len(T_position_counter[key])):
+                    dag = copy.deepcopy(original_dag)
```

Comment:
I think this could just be `dag = copy.deepcopy(dag)` i.e. only copying the dag once.

Reply:
I could be wrong, but at the moment I don't think that this would work. The `copy.deepcopy()` in line 72 is needed in every iteration, as we would otherwise loose the optimization of other phase-rotation gates made in previous runs of the `for key in T_counter:` loop. 

The `copy.deepcopy()` in line 74 is also needed in every iteration of `for final_gate_position in range(len(T_position_counter[key])):`, as the algorithm would not work otherwise. Right?

Reply:
I meant replacing the copy on line 74 and removing the copy on line 72. It seems a copy is made, and then the only use of it is to make another copy? I think that this could been made from the original dag, but I could be mistaken! It may be more clear if the variable on line 74  is declared as something other than `dag`, perhaps something like `current_dag`

Reply:
Thank you for pointing this out. We had a careful look at the code, and simply removing the copy in line 72 isn't possible. Nevertheless, we declared the `dag` in line 74 as the new variable `dag_copy`, which enabled us to further simplify the code.